### PR TITLE
Add rollout service in grpc api

### PR DESCRIPTION
**Pull request by hannesg**
Add the rollout GRPC service to the frontend code

image

TODO(HVG): write tests

### CODE REVIEW

File: services/frontend-service/src/ui/components/ReleaseCard/ReleaseCard.tsx [hunk 1]
```diff
@@ -15,25 +15,111 @@ along with kuberpult. If not, see <https://directory.fsf.org/wiki/License:Expat>
...
+            return status;
+        })
+    );
+    rolloutEnvs.sort((a, b) => {
+        if (a.environment < b.environment) {
+            return -1;
+        } else if (a.environment > b.environment) {
+            return 1;
+        }
+        return 0;
```
Comment (sven-urbanski-freiheit-com) [inline services/frontend-service/src/ui/components/ReleaseCard/ReleaseCard.tsx:104]: fine for now, but later on we should group by envGroup as well

File: services/frontend-service/src/ui/components/ReleaseCard/ReleaseCard.tsx [hunk 2]
```diff
@@ -15,25 +15,111 @@ along with kuberpult. If not, see <https://directory.fsf.org/wiki/License:Expat>
...
+            return <span className="rollout__description_successful">✓ Done</span>;
+        case RolloutStatus.RolloutStatusProgressing:
+            return <span className="rollout__description_progressing">↻ In progress</span>;
+        case RolloutStatus.RolloutStatusError:
+            return <span className="rollout__description_error">! Failed</span>;
+    }
+    return <span className="rollout__description_unknown">? Unkwown</span>;
+};
+
+const rolloutStatusPriority = [
```
Comment (sven-urbanski-freiheit-com) [inline services/frontend-service/src/ui/components/ReleaseCard/ReleaseCard.tsx:64]:
```suggestion
// note that the order is important here.
// "most interesting" must come first.
// see `calculateDeploymentStatus`
const rolloutStatusPriority = [
```

File: services/frontend-service/src/ui/components/ReleaseCard/ReleaseCard.tsx [hunk 3]
```diff
@@ -15,25 +15,111 @@ along with kuberpult. If not, see <https://directory.fsf.org/wiki/License:Expat>
...
+    RolloutStatus.RolloutStatusError,
+    RolloutStatus.RolloutStatusProgressing,
+    RolloutStatus.RolloutStatusUnknown,
```
Comment (sven-urbanski-freiheit-com) [inline services/frontend-service/src/ui/components/ReleaseCard/ReleaseCard.tsx:67]:
```suggestion
    RolloutStatus.RolloutStatusUnknown,
    RolloutStatus.RolloutStatusSuccessful,
```

File: services/frontend-service/src/ui/components/ReleaseCard/ReleaseCard.tsx [hunk 4]
```diff
@@ -54,8 +140,29 @@ export const ReleaseCard: React.FC<ReleaseCardProps> = (props) => {
                     <FormattedDate className={'date'} createdAt={createdAt} />
                 </div>
             )}
+            {rolloutEnvs.length > 0 && (
```
Comment (sven-urbanski-freiheit-com) [inline services/frontend-service/src/ui/components/ReleaseCard/ReleaseCard.tsx:143]: should we show some message if rolloutEnvs.length == 0?

File: services/frontend-service/src/ui/utils/store.tsx [hunk 5]
```diff
@@ -622,3 +623,36 @@ export const useConfigReady = (): boolean => useFrontendConfig(({ configReady })
...
+
+export const useRolloutStatus = (application: string): RolloutStatusApplication =>
+    useEntireRolloutStatus((data: RolloutStatusStore): RolloutStatusApplication => data[application] ?? {});
+
+export const UpdateRolloutStatus = (ev: StreamStatusResponse): void => {
+    rolloutStatus.set((data: RolloutStatusStore) => {
+        let app = data[ev.application];
+        if (!app) {
+            app = data[ev.application] = {};
+        }
```
Comment (sven-urbanski-freiheit-com) [inline services/frontend-service/src/ui/utils/store.tsx:645]:
```suggestion
        const app = data[ev.application] || {};
```

File: services/frontend-service/src/ui/utils/store.tsx [hunk 6]
```diff
@@ -622,3 +623,36 @@ export const useConfigReady = (): boolean => useFrontendConfig(({ configReady })
...
+        app[ev.environment] = ev;
+    });
+};
+
+export const FlushRolloutStatus = (): void => {
+    rolloutStatus.set((data: RolloutStatusStore) => {
+        const result: RolloutStatusStore = {};
```
Comment (sven-urbanski-freiheit-com) [inline services/frontend-service/src/ui/utils/store.tsx:652]: why can't we just return `{}` here?

File: services/frontend-service/src/ui/utils/store.tsx (hunk 5 above)
Comment (hannesg) [reply services/frontend-service/src/ui/utils/store.tsx:645]: This is not the same. My code also sets the value in the map.

File: services/frontend-service/src/ui/utils/store.tsx (hunk 6 above)
Comment (hannesg) [reply services/frontend-service/src/ui/utils/store.tsx:652]: I also thought that but actually we can't. It turned out that the lib we are using is doing sparse updates of the store.

File: services/frontend-service/src/ui/components/ReleaseCard/ReleaseCard.tsx (hunk 1 above)
Comment (hannesg) [reply services/frontend-service/src/ui/components/ReleaseCard/ReleaseCard.tsx:108]: Maybe yes, on the other hand, it's also interesting to see it for each cluster. What do you think we should do?

Comment (sven-urbanski-freiheit-com) [reply services/frontend-service/src/ui/components/ReleaseCard/ReleaseCard.tsx:108]: yes, show for each cluster, but sort/group-by env group.

File: services/frontend-service/src/ui/utils/store.tsx (hunk 5 above)
Comment (sven-urbanski-freiheit-com) [reply services/frontend-service/src/ui/utils/store.tsx:645]: ahh, you're right

**Pull request merged**